File: packages-gen/dart/libs/xt-substrate/lib/base-json.dart
import 'dart:convert';

frame_kindp(kind) {
  return (kind == "request") || (kind == "response") || (kind == "stream") || (kind == "subscribe") || (kind == "unsubscribe");
}

valid_framep(value) {
  if(!(("Map" == (value.runtimeType).toString()) || (value.runtimeType).toString().startsWith("_Map") || (value.runtimeType).toString().startsWith("LinkedMap"))){
    return false;
  }
  var kind = value["kind"];
  var id = value["id"];
  var space = value["space"];
  var meta = value["meta"];
  if(!(() {
    var dart_truthy__51343 = frame_kindp(kind);
    return (null != dart_truthy__51343) && (false != dart_truthy__51343);
  })() || !("String" == (id.runtimeType).toString()) || !("String" == (space.runtimeType).toString()) || ((null != meta) && !(("Map" == (meta.runtimeType).toString()) || (meta.runtimeType).toString().startsWith("_Map") || (meta.runtimeType).toString().startsWith("LinkedMap")))){
    return false;
  }
  if(kind == "request"){
    return ("String" == (value["action"].runtimeType).toString()) && ((null == value["args"]) || ((value["args"].runtimeType).toString().startsWith("List") || (value["args"].runtimeType).toString().startsWith("_GrowableList")));
  }
  else if(kind == "response"){
    return ("String" == (value["reply_to"].runtimeType).toString()) && ("String" == (value["status"].runtimeType).toString());
  }
  else{
    return "String" == (value["signal"].runtimeType).toString();
  }
}

normalize_error(err) {
  if(null == err){
    return null;
  }
  else if("String" == (err.runtimeType).toString()){
    return <dynamic, dynamic>{"message":err};
  }
  else if(("Map" == (err.runtimeType).toString()) || (err.runtimeType).toString().startsWith("_Map") || (err.runtimeType).toString().startsWith("LinkedMap")){
    var out = xt.lang.common_data.obj_clone(err);
    if(!out.containsKey("message") && out.containsKey("error") && ("String" == (out["error"].runtimeType).toString())){
      out["message"] = out["error"];
    }
    if(!out.containsKey("message") && out.containsKey("status") && ("String" == (out["status"].runtimeType).toString())){
      out["message"] = out["status"];
    }
    if(!out.containsKey("message")){
      out["message"] = (err).toString();
    }
    return out;
  }
  else{
    return <dynamic, dynamic>{"message":(err).toString()};
  }
}

normalize_frame(frame) {
  if(!(("Map" == (frame.runtimeType).toString()) || (frame.runtimeType).toString().startsWith("_Map") || (frame.runtimeType).toString().startsWith("LinkedMap"))){
    return frame;
  }
  var out = xt.lang.common_data.obj_clone(frame);
  if(("response" == out["kind"]) && ("error" == out["status"]) && (null != out["error"])){
    out["error"] = normalize_error(out["error"]);
  }
  return out;
}

encode_frame(frame) {
  var out = normalize_frame(frame);
  if(!(() {
    var dart_truthy__51344 = valid_framep(out);
    return (null != dart_truthy__51344) && (false != dart_truthy__51344);
  })()){
    throw "invalid node json frame";
  }
  return jsonEncode(out);
}

decode_frame(input) {
  var out = ("String" == (input.runtimeType).toString()) ? jsonDecode(input) : input;
  if(!(() {
    var dart_truthy__51345 = valid_framep(out);
    return (null != dart_truthy__51345) && (false != dart_truthy__51345);
  })()){
    throw "invalid node json frame";
  }
  return out;
}

File: packages-gen/js/libs/xt-substrate/base-json.js
function frame_kindp(kind){
  return (kind == "request") || (kind == "response") || (kind == "stream") || (kind == "subscribe") || (kind == "unsubscribe");
}

function valid_framep(value){
  if(!((null != value) && ("object" == (typeof value)) && !Array.isArray(value))){
    return false;
  }
  let kind = value["kind"];
  let id = value["id"];
  let space = value["space"];
  let meta = value["meta"];
  if(!frame_kindp(kind) || !("string" == (typeof id)) || !("string" == (typeof space)) || ((null != meta) && !((null != meta) && ("object" == (typeof meta)) && !Array.isArray(meta)))){
    return false;
  }
  if(kind == "request"){
    return ("string" == (typeof value["action"])) && ((null == value["args"]) || Array.isArray(value["args"]));
  }
  else if(kind == "response"){
    return ("string" == (typeof value["reply_to"])) && ("string" == (typeof value["status"]));
  }
  else{
    return "string" == (typeof value["signal"]);
  }
}

function normalize_error(err){
  if(null == err){
    return null;
  }
  else if("string" == (typeof err)){
    return {"message":err};
  }
  else if((null != err) && ("object" == (typeof err)) && !Array.isArray(err)){
    let out = Object.assign({},err);
    if(!(null != out["message"]) && (null != out["error"]) && ("string" == (typeof out["error"]))){
      out["message"] = out["error"];
    }
    if(!(null != out["message"]) && (null != out["status"]) && ("string" == (typeof out["status"]))){
      out["message"] = out["status"];
    }
    if(!(null != out["message"])){
      out["message"] = String(err);
    }
    return out;
  }
  else{
    return {"message":String(err)};
  }
}

function normalize_frame(frame){
  if(!((null != frame) && ("object" == (typeof frame)) && !Array.isArray(frame))){
    return frame;
  }
  let out = Object.assign({},frame);
  if(("response" == out["kind"]) && ("error" == out["status"]) && (null != out["error"])){
    out["error"] = normalize_error(out["error"]);
  }
  return out;
}

function encode_frame(frame){
  let out = normalize_frame(frame);
  if(!valid_framep(out)){
    throw "invalid node json frame";
  }
  return JSON.stringify(out);
}

function decode_frame(input){
  let out = ("string" == (typeof input)) ? JSON.parse(input) : input;
  if(!valid_framep(out)){
    throw "invalid node json frame";
  }
  return out;
}

module.exports = {
  ["frame_kindp"]:frame_kindp,
  ["valid_framep"]:valid_framep,
  ["normalize_error"]:normalize_error,
  ["normalize_frame"]:normalize_frame,
  ["encode_frame"]:encode_frame,
  ["decode_frame"]:decode_frame
}